cpp/include/cudf/contiguous_split.hpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace cudf {
+
+/**
+ * @addtogroup column_copy
+ * @{
+ * @file
+ * @brief Table APIs for contiguous_split, pack, unpack, and metadadata
+ */
+
+/**
+ * @brief Column data in a serialized format
+ *
+ * @ingroup copy_split
+ *
+ * Contains data from an array of columns in two contiguous buffers: one on host, which contains
+ * table metadata and one on device which contains the table data.
+ */
+struct packed_columns {
+  /**
+   * @brief Host-side metadata buffer used for reconstructing columns via unpack.
+   *
+   * @ingroup copy_split
+   */
+  struct metadata {
+    metadata() = default;
+
+    /**
+     * @brief Construct a new metadata object
+     *
+     * @param v Host-side buffer containing metadata
+     */
+    metadata(std::vector<uint8_t>&& v) : data_(std::move(v)) {}
+
+    /**
+     * @brief Returns pointer to the host-side metadata buffer data
+     *
+     * @return Pointer to the host-side metadata buffer
+     */
+    [[nodiscard]] uint8_t const* data() const { return data_.data(); }
+
+    /**
+     * @brief Returns size of the metadata buffer
+     *
+     * @return Size of the metadata buffer
+     */
+    [[nodiscard]] size_t size() const { return data_.size(); }
+
+   private:
+    std::vector<uint8_t> data_;
+  };
+
+  packed_columns()
+    : metadata_(std::make_unique<metadata>()), gpu_data(std::make_unique<rmm::device_buffer>())
+  {
+  }
+
+  /**
+   * @brief Construct a new packed columns object
+   *
+   * @param md Host-side metadata buffer
+   * @param gd Device-side data buffer
+   */
+  packed_columns(std::unique_ptr<metadata>&& md, std::unique_ptr<rmm::device_buffer>&& gd)
+    : metadata_(std::move(md)), gpu_data(std::move(gd))
+  {
+  }
+
+  std::unique_ptr<metadata> metadata_;           ///< Host-side metadata buffer
```

Comment:
```suggestion
  std::vector<uint8_t> metadata_;           ///< Host-side metadata buffer
```

Reply:
Also, recommend to use `std::vector<std::byte>` instead.

Reply:
Yeah I just moved this code, but I see the point. I'll push a commit to remove the struct in favor of a `using`, which means the rest of the code that wants to use `metadata` as a type can still do so.

Reply:
@ttnghia you marked this as resolved, but I haven't pushed it. Do you want me stick with `uint8_t`? I have been making the change locally and the main thing is I have to change other files in java/python and other source files on the cpp side. If we are OK with the bigger diff, I can spend time on moving us to this.

Reply:
Oh sorry, accidentally marked as resolved :smile: . Yes, please address this.

Reply:
I have tried this, but it doesn't seem `std::byte` is defined in cpython:

```
cdef extern from "cudf/contiguous_split.hpp" namespace \
        "cudf::packed_columns" nogil:
    ctypedef vector[byte] metadata 
                       ^
------------------------------------------------------------

cudf/python/cudf/cudf/_lib/cpp/contiguous_split.pxd:16:24: unknown type in template argument
```

Looking at `stddef` in cython, I see they didn't exposed that type: https://github.com/cython/cython/blob/1aef71764cd50e4cd2f638eac8e69fff3ced2224/Cython/Includes/libc/stddef.pxd

Should we leave as uint8_t? Or should I handle this differently? @ttnghia

Reply:
@ttnghia I did remove the typedef for `metadata` and went directly with the `std::vector<uint8_t>` here: https://github.com/rapidsai/cudf/pull/13186/commits/d8dd74683e6e86f4de62456ae835720286bdd074

Reply:
Keeping `uint8_t` is fine. I want to move forward with a more standard way that libcudf devs prefer (`std::byte` instead of other types) but understand that it is may be out of scope of this PR.